src/openforms/emails/service.py

```diff
@@ -0,0 +1,56 @@
+from django_yubin.models import Message
+
+from openforms.submissions.models import Submission
+
+from .constants import (
+    X_OF_CONTENT_TYPE_HEADER,
+    X_OF_CONTENT_UUID_HEADER,
+    X_OF_EVENT_HEADER,
+    EmailContentTypeChoices,
+    EmailEventChoices,
+)
+
+__all__ = ["get_last_confirmation_email"]
+
+
+def get_last_confirmation_email(submission: Submission) -> tuple[str, int] | None:
+    """
+    Get the last created confirmation email from the
+    :class:`django_yubin.models.Message` model.
+
+    :return: A tuple of the HTML content and message ID of the last created confirmation
+      email for the specified submission. ``None`` if there weren't any.
+    """
+    # We only care about messages that were created after the submission
+    messages = Message.objects.filter(date_created__gte=submission.created_on).order_by(
+        "-date_created"
+    )
+
+    for message in messages:
+        email = message.get_email_message()
```

Comment:
we can narrow down the queryset more before checking in python code by filtering on the `_message_data` text field which contains the raw e-mail message.

```suggestion
    messages = Message.objects.filter(
        date_created__gte=submission.created_on,
        _message_data__contains=X_OF_CONTENT_TYPE_HEADER,
        _message_data__contains=EmailContentTypeChoices.submission,
        _message_data__contains=str(submission.uuid),
    ).order_by(
        "-date_created"
    )

    for message in messages:
        email = message.get_email_message()
```
this should substantially narrow down the result set.

It's probably a good idea to test this on a big environment with many email logs to see the performance of the query.

Reply:
Would there be a way to do that before releasing it first?

Reply:
Let me see if I can get my hands on a dataset tomorrow!

Reply:
So - the conclusion here for me would be:

* Add a migration in our own `emails` app to add an index to django yubins `Message` model. You can use `SeparateDatabaseAndState` for this.
* Add the submission UUID query/filter to the queryset, but not more than that. The other fluff seems to make query performance worse
* The high specificity should result in very few messages being loaded into memory to perform the remaining header checks. I'd use a `.iterator()` on the queryset though to be sure we don't load too much data in memory, since those e-mail messages can contain attachments as well.

Reply:
Thank you :bow:

I will have a look!